### ISSUE TITLE
Provide way of forcing a 404

### ISSUE THREAD

**Issue by gryphonmyers** [brillout/vite-plugin-ssr#103]
Let's say you have a page file that matches the current route, but upon fetching some data in the addPageContext hook, you determine that necessary data is unavailable and the user should be served with a 404 error page. At present, I believe the only option available to a user here is to halt rendering of the page and render an error would be to throw an error within the addPageContext hook. However, this will end up as a 500, not a 404. In this specific case, we are deliberately throwing an error that we would like to be cast as a 404. 

I have a suggestion for how this could be achieved - export error classes that can be imported and used to throw specific error types. E.g.

```javascript

import { ServerError, NotFoundError } from 'vite-plugin-ssr';

export async function addPageContext({ url }) {

  const criticalData = await fetchCriticalData(url);

  if (!criticalData) {
    throw new NotFoundError;
  }
  return {
    criticalData
  }
}


```

**Comment by gryphonmyers** [comment]
Alternatively, you could take a more hands off approach and make it clear that error handling is totally up to the user, and allow them to handle it however they want in their render function. That honestly might be better - backing off the whole concept of "is404" and instead just have 404s be an error class used internally that can be easily identified within the user's error handling logic

**Comment by brillout** [comment]
What is your concrete use case?

You can already hack something like this today.

```js
export async function render() {
  if (!criticalData) {
    return {dataMissing: true};
  }
}
```

```js
// server.js

const renderPage = createPageRender(/*...*/)

app.get('*', async (req, res, next) => {
  const result = await renderPage({ url: req.url })
  if (result.dataMissing) {
    const result = await renderPage({ url: '/doesNotExist' })
    res.status(result.statusCode).send(result.renderResult)
    return;
  }
})
```

>  return render404Error(Page)

Yea I was thinking about something like this as well.

**Comment by gryphonmyers** [comment]
> What is your concrete use case?

My concrete use case is the existence of a particular page file being dependent on the existence of data fetched from a database (a "page" model). In some instances this model may not exist, and if it doesn't, we would like to handle the render as if it were a 404. I was able to achieve this by doing something similar to my example (throw an error with `{ status: 404 }` assigned to it, then inspect the error in _default.server.js). This works perfectly fine, the issue is mostly about providing clear guidance about a convention for error handling. Also, since this error is in the pageContext as `_err` it does feel "wrong" to be handling it since the underscore prefix could indicate it's an internal API and not meant for public consumption / subject to change.

**Comment by brillout** [comment]
How about:

```js
// .page.server.js

export async function render() {
  if (!criticalData) {
    return {dataMissing: true};
  }
}
```

```js
 // server.js

const { renderErrorPage } = createPageRender(/*...*/)

app.get('*', async (req, res, next) => {
  const { url } = req.url
  let result = await renderPage({ url })
  if (result.dataMissing) {
    result = await renderErrorPage({ url, is404: true })
  }
  res.status(result.statusCode).send(result.renderResult)
})
```

Concerns about this?

> issue is mostly about providing clear guidance about a convention for error handling

I agree; I'll add docs about this.

**Comment by gryphonmyers** [comment]
> How about:
> 
> ```js
> // .page.server.js
> 
> export async function render() {
>   if (!criticalData) {
>     return {dataMissing: true};
>   }
> }
> ```
> 
> ```js
>  // server.js
> 
> const { renderErrorPage } = createPageRender(/*...*/)
> 
> app.get('*', async (req, res, next) => {
>   const { url } = req.url
>   let result = await renderPage({ url })
>   if (result.dataMissing) {
>     result = await renderErrorPage({ url, is404: true })
>   }
>   res.status(result.statusCode).send(result.renderResult)
> })
> ```
> 
> Concerns about this?
> 
> > issue is mostly about providing clear guidance about a convention for error handling
> 
> I agree; I'll add docs about this.

Do you have reservations about making thrown errors the single path into the error template? Then errors can all be handled the same way and we don't have to make any exceptions or set any special state for 404. To me that sounds more versatile as it would accomplish my use case as well as any number of other error handling requirements. For example, my use case is "data missing" but what if someone is querying a backend service that times out and they want to render a special error message for this? With my suggestion, they could use whatever mechanism they like for identifying errors (perhaps https://www.npmjs.com/package/http-errors) and using whatever rendering logic they want based on the type of error identified. 

Again, this is already possible, it just feels weird using a property called _err. I'm basically suggesting making the error accessible under a non-prefixed key. You could also consider passing the error as an additional argument to the render function instead of putting it in context if you're worried about polluting the namespace.

**Comment by brillout** [comment]
The idea of my solution is that I want users to get used to the fact that in `const result = await renderPage()` the `result` value is simply what is returned by the `render()` hook. It's a cornerstone design about vite-plugin-ssr that I want my users to get familiar with.

If you expect an exception to be able to throw then you can catch it at the origin and propagate the information using idiomatic vite-plugin-ssr code.

In short: you don't need to access the private property `pageContext._err` with my proposed solution.

**Comment by gryphonmyers** [comment]
Yeah but the downside is your error handling then occurs outside of the rendering logic, even though the _error page file is part of the application (and already available in the render scope as the current Page). The render function seems like a decent place to be doing the error handling because all the information is already there. How would you event instruct the render to use the _error page template if you are explicitly running the render again? Also how would this work in a pre-rendering context? If I return `dataMissing: true` will it know to skip those routes?

I also just realized you were proposing returning `dataMissing: true` in the render hook not the `addPageContext` hook. Can you call the _default render function from page render functions? (Very similar to conversations we've had elsewhere about overriding the addPageContext hook).

**Comment by brillout** [comment]
Let me clarify:

```js
// some.page.server.js

export async function addPageContext({ url }) {
  const criticalData = await fetchCriticalData(url);

  if (!criticalData) {
    return { dataIsMissing: true }
  }
  
  return { criticalData }
}
```

```js
// _default.page.server.js

export function render({ dataIsMissing, Page }) {
  if( dataIsMissing ) {
    return { dataIsMissing: true };
  }
}
```

```js
 // server.js

const { renderPage, renderErrorPage } = createPageRender(/*...*/)

app.get('*', async (req, res, next) => {
  const { url } = req.url
  let result = await renderPage({ url })
  if (result.dataMissing) {
    // We can add additional `pageContext` so that the error page knows that data was missing.
    result = await renderErrorPage({ url, is404: true, dataIsMissing: true })
  }
  res.status(result.statusCode).send(result.renderResult)
})
```

> How would you event instruct the render to use the _error page template if you are explicitly running the render again?

We are not calling `renderPage` but `renderErrorPage`.

And note that, as usual, you can pass any initial pageContext to `renderErrorPage`. So that you can let the error page know that is has been called because some data were missing.

> Also how would this work in a pre-rendering context?

While pre-rendering, you wouldn't have the problem in the first place because you provide the list of URLs that are to be pre-rendered and you wouldn't provide a URL that doesn't exist (since the list of URLs will come from your source of truth).

**Comment by gryphonmyers** [comment]
> Let me clarify:
> 
> ```js
> // some.page.server.js
> 
> export async function addPageContext({ url }) {
>   const criticalData = await fetchCriticalData(url);
> 
>   if (!criticalData) {
>     return { dataIsMissing: true }
>   }
>   
>   return { criticalData }
> }
> ```
> 
> ```js
> // _default.page.server.js
> 
> export function render({ dataIsMissing, Page }) {
>   if( dataIsMissing ) {
>     return { dataIsMissing: true };
>   }
> }
> ```
> 
> ```js
>  // server.js
> 
> const { renderPage, renderErrorPage } = createPageRender(/*...*/)
> 
> app.get('*', async (req, res, next) => {
>   const { url } = req.url
>   let result = await renderPage({ url })
>   if (result.dataMissing) {
>     // We can add additional `pageContext` so that the error page knows that data was missing.
>     result = await renderErrorPage({ url, is404: true, dataIsMissing: true })
>   }
>   res.status(result.statusCode).send(result.renderResult)
> })
> ```
> 
> > How would you event instruct the render to use the _error page template if you are explicitly running the render again?
> 
> We are not calling `renderPage` but `renderErrorPage`.

Oh I didn't realize renderErrorPage would be a built-in function. I thought you were saying the burden of rendering an error would be totally on the user. 

> And note that, as usual, you can pass any initial pageContext to `renderErrorPage`. So that you can let the error page know that is has been called because some data were missing.
> 
> > Also how would this work in a pre-rendering context?
> 
> While pre-rendering, you wouldn't have the problem in the first place because you provide the list of URLs that are to be pre-rendered and you wouldn't provide a URL that doesn't exist (since the list of URLs will come from your source of truth).

Yeah ok duh I just wasn't thinking clearly when I wrote this :D We should still consider however that the pre-render process does generate a 404 placeholder - is this something that the we should give the user more control over? For example, what if they want to pre-render not only a 404 but a 500, 422, whatever else they want, which they could then map to e.g. via Cloudfront error handling?

**Comment by brillout** [comment]
I also often get confused; it's not like SSR is the easiest piece of technology out there :-).

I believe, but correct me if I'm wrong, that after pre-rendering, there can only be a status code of `200`. (At least as you as you don't do a partial one `$ vite-plugin-ssr prerender --partial`.)

Btw I love that you try to forecast possible scenarios. I also try to do that, which is crucially important when designing an architecture 👌.

(Btw. I'm still regularly thinking about our conversation on how we should design server hooks.)

**Comment by gryphonmyers** [comment]
> I also often get confused; it's not like SSR is the easiest piece of technology out there :-).
> 
> I believe, but correct me if I'm wrong, that after pre-rendering, there can only be a status code of `200`. (At least as you as you don't do a partial one `$ vite-plugin-ssr prerender --partial`.)
> 
> Btw I love that you try to forecast possible scenarios. I also try to do that, which is crucially important when designing an architecture 👌.
> 
> (Btw. I'm still regularly thinking about our conversation on how we should design server hooks.)

The setup I'm describing is more of a static/dynamic hybrid using edge functions. For instance, AWS Cloudfront allows you to handle errors from your origin by mapping specific error codes to specific html files. With Lambda@Edge handlers in place, one can easily introduce non-404 errors (e.g. 503, 504), so this would give the user a way of mapping those errors to static pre-rendered files. Then the user will be served with a pretty error rather than a stock AWS one, without the added complexity of having to render that error within the Lambda@Edge handler.

**Comment by brillout** [comment]
I see. Maybe creating normal pages "503.page.js", "504.page.js", etc. and then map `/503` and `/504` would do the trick? I guess the status code would be handled by AWS then.

**Comment by gryphonmyers** [comment]
I guess my thought is, we are already generating a "fake" 404 page - what is special about 404? Why not allow the user to choose  which errors get rendered in this way? Something like `vite-plugin-ssr prerender --render-error 404 --render-error 403` or comma-separated values if you'd prefer

**Comment by brillout** [comment]
Static host always need a `404.html`. Wanting other types of error to be pre-rendered is quite an edge case, so I'm not sure we want add options for something that can be already achieved by other means. anyways To avoid feature creep.

**Comment by brillout** [comment]
I still struggle to see a use case for this. But I'm happy to be shown wrong and more than happy to re-open this.
